examples: How to launch and reattach to a long-running process

martinpitt opened the pull request:
Cockpit pages should never spawn long-running precious processes, as a
cockpit session is not reliably connected for a long time:  Network
interfaces go down or  roam, TCP timeouts happen, browsers crash, tabs
get closed accidentally, batteries drain, people close/suspend their
laptops, and so on. 

Most services that Cockpit talks to manage their own runtime state and
jobs, such as udisks, libvirt, or podman. But this is not the case for
e.g. running Ansible playbooks or launching installer scripts. These
should be wrapped into a transient service unit, so that systemd takes
over the role of the job manager.

Add a "long-running-process" example which demonstrates this. It runs an
arbitary shell command in a transient cockpit-longrunning-$USER.service,
shows the live log output, and reattaches to it on page load.

The most important design question is the choice of the service name,
especially if the page can manage more than one process (such as
launching parallel playbooks with different environment variables). It
must contain exactly all the identifying properties in the name, so that
it is predictable.

martinpitt commented:
[fails on rhel-8-3](https://logs.cockpit-project.org/logs/pull-14318-20200706-102459-cd7ee461-rhel-8-3/log.html#301-1), same on -distropkg, so not a flake.

mvollmer commented:
This is nice! My immediate reaction is that, yes, we should offer this as some library.

martinpitt commented:
There was a race condition. Fixed now, although in a slightly hackish way. But let's at least get to a known-good place.

martinpitt commented:
This works, but I'm not happy with it yet:

 - [ ] On second thought, the user name doesn't belong into the service name. A different admin should see the same thing, as it's being run as root. And for an user operation, these run on different user systemds anyway. So drop that.
 - [ ] Replace the polling with Job signals.
 - [ ] Once we have signals, react to the unit finishing as well.
 - [ ] See how much makes sense to be put into a library. At first sight, pkg/lib/service.js does not have much useful stuff for this use case, but some of it may generalize into its own little library.
 - [ ] Add a README to explain the purpose and strategy.

martinpitt commented:
Reworked using `Job{New,Removed}` signals, which is much nicer. I also added a README.

I haven't yet looked into how much of it is generic and can be factored out. At first sight, `checkState()`, `run()`, and the JobNew/Removed handling of `setUp()` are generic. The journal bits are not, it is too much entangled with how the page looks like and what it wants to do. So checkState() needs some abstraction/indirection/signal how to tell the main app to start log tailing.

martinpitt commented:
Still fails [on](https://logs.cockpit-project.org/logs/pull-14318-20200708-164457-41961106-ubuntu-2004/log.html#301-1) [Ubuntu](https://logs-https-cockpit.apps.ci.centos.org/logs/pull-14318-20200708-164457-41961106-ubuntu-stable/log.html)

martinpitt commented:
I believe this is sufficiently robust and race free now as a general approach, so time for a first review. @mvollmer, do you have time to have a look? @allisonkarlitskaya , you may also be interested.

I'm not sure if we want to block this on moving out the approach into a library, or do this as a follow-up. My gut feeling is the latter, but comments appreciated.

mvollmer commented:
> I'm not sure if we want to block this on moving out the approach into a library, or do this as a follow-up. My gut feeling is the latter, but comments appreciated.

I agree.  Let's try to find a user in Cockpit itself, maybe installing a new VM?  Running sosreport?

martinpitt commented:
sosreport sounds like a great thing to convert. Then we can better see how much code can be shared, and how to generalize it, and then factor it out. But at least ovirt and other teams then already have something to look at.

martinpitt commented:
@allisonkarlitskaya : Indeed, see above for the discussion -- this should become a library, we just need a first real-world use case for that. And we identified one.

Very good point about the user; right now it runs as root, but that  should be documented in a comment.

martinpitt commented:
@allisonkarlitskaya : I added explanations about "runs as root" both to the JS as comment and to the README.

martinpitt commented:
Actually, that was bogus. You *can* run systemd-run for user jobs, with `--user`. Adjusted accordingly.

allisonkarlitskaya commented on examples/long-running-process/long-running.js:
```diff
@@ -0,0 +1,125 @@
+
+/* Start long-running process, called on clicking the "Start Process" button.
+ * This runs as root, thus will be shared with all privileged Cockpit sessions. If the process
+ * should run with user privileges instead, drop the `superuser` and call `systemd-run` with `--user`.
```
It's more complicated than "just add --user" here.  The journal command would need to be adjusted, and (probably) also the systemd1 instance you connect to.

It's fine to just say "this runs stuff as root" and not contemplate the user possibility.  If you do want to talk about the user one, it should probably get a more complete treatment.

martinpitt replied:
Indeed, that's incomplete. It would change quite a lot of code. If we ever get this use case (running as root is the much more common one), I'd rather suggest to add a completely separate example with its own test. By then we also have a library for this.

I entirely removed that comment.
